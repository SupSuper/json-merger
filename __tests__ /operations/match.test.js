const jsonMerger = require("../../dist");
const {testConfig} = require("../__helpers__");

describe("when merging two arrays and a source array item has a $match indicator", function () {

    test("and $match.index is set it should search for a match in the target array and use $match.then as value", function () {

        const object1 = {
            "a": [
                1,
                2,
                3
            ]
        };

        const object2 = {
            "a": [
                {
                    "$match": {
                        "index": 1,
                        "then": {
                            "$remove": true
                        }
                    }
                }
            ]
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $match.index is set to '-' it should match with the last item in the target array and use $match.then as value", function () {

        const object1 = {
            "a": [
                1,
                2,
                3
            ]
        };

        const object2 = {
            "a": [
                {
                    "$match": {
                        "index": "-",
                        "then": {
                            "$remove": true
                        }
                    }
                }
            ]
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $match.query is set it should search for a match in the target array and use $match.then as value", function () {

        const object1 = {
            "a": [
                1,
                2,
                3
            ]
        };

        const object2 = {
            "a": [
                {
                    "$match": {
                        "query": "$[1]",
                        "then": {
                            "$remove": true
                        }
                    }
                }
            ]
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $match.path is set it should search for a match in the target array and use $match.then as value", function () {

        const object1 = {
            "a": [
                1,
                2,
                3
            ]
        };

        const object2 = {
            "a": [
                {
                    "$match": {
                        "path": "/1",
                        "then": {
                            "$remove": true
                        }
                    }
                }
            ]
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("it should ignore the item if no matching $match.index has been found", function () {

        const object1 = {
            "a": [
                1,
                2,
                3
            ]
        };

        const object2 = {
            "a": [
                {
                    "$match": {
                        "index": 99,
                        "then": {
                            "$remove": true
                        }
                    }
                }
            ]
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("it should ignore the item if no matching $match.query has been found", function () {

        const object1 = {
            "a": [
                1,
                2,
                3
            ]
        };

        const object2 = {
            "a": [
                {
                    "$match": {
                        "query": "$[99]",
                        "then": {
                            "$remove": true
                        }
                    }
                }
            ]
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("it should process and merge the $match.then value if a match is found", function () {

        const object1 = {
            "a": [
                {
                    "a1": "original"
                },
                {
                    "a2": "original"
                }
            ]
        };

        const object2 = {
            "a": [
                {
                    "$match": {
                        "index": 1,
                        "then": {
                            "$comment": "Should be stripped",
                            "b2": "merged"
                        }
                    }
                }
            ]
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });
});
