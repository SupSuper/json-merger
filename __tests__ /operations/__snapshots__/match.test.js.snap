// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when merging two arrays and a source array item has a $match indicator and $match.index is set it should search for a match in the target array and use $match.then as value 1`] = `
"{
	\\"a\\": [
		1,
		3
	]
}"
`;

exports[`when merging two arrays and a source array item has a $match indicator and $match.index is set to '-' it should match with the last item in the target array and use $match.then as value 1`] = `
"{
	\\"a\\": [
		1,
		2
	]
}"
`;

exports[`when merging two arrays and a source array item has a $match indicator and $match.path is set it should search for a match in the target array and use $match.then as value 1`] = `
"{
	\\"a\\": [
		1,
		3
	]
}"
`;

exports[`when merging two arrays and a source array item has a $match indicator and $match.pointer is set it should search for a match in the target array and use $match.then as value 1`] = `
"{
	\\"a\\": [
		1,
		3
	]
}"
`;

exports[`when merging two arrays and a source array item has a $match indicator it should ignore the item if no matching $match.index has been found 1`] = `
"{
	\\"a\\": [
		1,
		2,
		3
	]
}"
`;

exports[`when merging two arrays and a source array item has a $match indicator it should ignore the item if no matching $match.path has been found 1`] = `
"{
	\\"a\\": [
		1,
		2,
		3
	]
}"
`;

exports[`when merging two arrays and a source array item has a $match indicator it should process and merge the $match.then value if a match is found 1`] = `
"{
	\\"a\\": [
		{
			\\"a1\\": \\"original\\"
		},
		{
			\\"a2\\": \\"original\\",
			\\"b2\\": \\"merged\\"
		}
	]
}"
`;
